cirq-rigetti/cirq_rigetti/aspen_device_test.py

```diff
@@ -0,0 +1,181 @@
+    assert qubit2.x == 1 + sqrt(2)
+    assert qubit2.y == 1 / sqrt(2)
+    assert qubit2.z == 0
+
+    invalid_qubit = OctagonalQubit(0)
+    invalid_qubit.octagon_position = 9
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.x
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.y
+
+    assert np.isclose(qubit0.distance(OctagonalQubit(1)), 1)
+    assert qubit0.distance(OctagonalQubit(7)) == 1
+    with pytest.raises(TypeError):
+        _ = qubit0.distance(AspenQubit(0, 0))
+
+    with pytest.raises(ValueError):
+        _ = OctagonalQubit(8)
+
+
+def test_aspen_qubit():
+    """test AspenQubit properties and transformations"""
+    qubit10 = AspenQubit(1, 0)
+    assert qubit10.octagon == 1
+    assert qubit10.octagon_position == 0
+    assert qubit10.index == 10
+    assert qubit10.dimension == 2
+
+    assert qubit10.x == 3 + 3 / sqrt(2)
+    assert qubit10.y == 1 + sqrt(2)
+    assert np.isclose(qubit10.distance(AspenQubit(0, 7)), 3 + 2 / sqrt(2))
+    assert np.isclose(qubit10.distance(AspenQubit(1, 3)), 1 + sqrt(2))
+
+    qubit15 = AspenQubit(1, 5)
+    assert qubit15.x == (2 + sqrt(2))
+
+    qubit15 = AspenQubit(1, 1)
+    assert qubit15.x == (2 + sqrt(2)) + 1 + sqrt(2)
+
+    invalid_qubit = AspenQubit(0, 0)
+    invalid_qubit.octagon_position = 9
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.x
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.y
+
+    with pytest.raises(TypeError):
+        _ = qubit10.distance(OctagonalQubit(0))
+
+    with pytest.raises(ValueError):
+        _ = AspenQubit(1, 9)
+
+    assert qubit10.to_named_qubit() == cirq.NamedQubit('10')
+    assert AspenQubit.from_named_qubit(cirq.NamedQubit('10')) == AspenQubit(1, 0)
+    with pytest.raises(ValueError):
+        _ = AspenQubit.from_named_qubit(cirq.NamedQubit('s'))
+    with pytest.raises(ValueError):
+        _ = AspenQubit.from_named_qubit(cirq.NamedQubit('19'))
+
+    with pytest.raises(ValueError):
+        _ = qubit10.to_grid_qubit()
+    assert AspenQubit(0, 2).to_grid_qubit() == cirq.GridQubit(0, 0)
+    assert AspenQubit(0, 1).to_grid_qubit() == cirq.GridQubit(1, 0)
+    assert AspenQubit(1, 5).to_grid_qubit() == cirq.GridQubit(0, 1)
+    assert AspenQubit(1, 6).to_grid_qubit() == cirq.GridQubit(1, 1)
+    assert AspenQubit.from_grid_qubit(cirq.GridQubit(1, 1)) == AspenQubit(1, 6)
+    with pytest.raises(ValueError):
+        _ = AspenQubit.from_grid_qubit(cirq.GridQubit(3, 4))
+
+
+def test_rigetti_qcs_aspen_device(qcs_aspen8_isa: InstructionSetArchitecture):
+    """test RigettiQCSAspenDevice validations"""
+    device = RigettiQCSAspenDevice(isa=qcs_aspen8_isa)
+
+    assert len(device.qubit_topology.nodes) == 32
+    assert len(device.qubit_topology.edges) == 38
+
+    valid_qubits = [
+        cirq.GridQubit(0, 0),
+        cirq.GridQubit(1, 1),
+        cirq.LineQubit(30),
+        cirq.NamedQubit('33'),
+        AspenQubit(3, 6),
+        OctagonalQubit(6),
+    ]
```

Comment:
This and the invalid qubits also seem like good candidates for parameterized tests.